rir/src/compiler/opt/scope_resolution.cpp

```diff
@@ -70,52 +72,74 @@ class TheScopeResolution {
                         ld->env(aload.env);
-                    } else {
-                        aval.eachSource([&](ValOrig& src) {
-                            // inter-procedural scope analysis can drag in
-                            // values
-                            // from other functions, which we cannot use here!
-                            if (src.origin->bb()->owner != function) {
-                                localVals = false;
+                    } else if (onlyLocalVals) {
+                        auto replaceLdFun = [&](Value* val) {
+                            if (val->type.isA(PirType::closure())) {
+                                next = bb->remove(ip);
+                                ld->replaceUsesWith(val);
+                                return;
+                            }
+                            auto fz = new Force(val);
+                            auto ch = new ChkClosure(fz);
+                            bb->replace(ip, fz);
+                            next = bb->insert(ip + 1, ch);
+                            next++;
+                            ld->replaceUsesWith(ch);
+                        };
+                        // This load can be resolved to a unique value
+                        aval.ifSingleValue([&](Value* val) {
+                            if (ldfun) {
+                                replaceLdFun(val);
+                            } else {
+                                ld->replaceUsesWith(val);
+                                next = bb->remove(ip);
                             }
                         });
-                        if (localVals) {
-                            aval.ifSingleValue([&](Value* val) {
-                                if (ldf) {
-                                    auto f = new Force(val);
-                                    // TODO
-                                    // !(PirType(RType::closure) >= phi->type)
-                                    ld->replaceUsesWith(f);
-                                    bb->replace(ip, f);
-                                } else {
-                                    ld->replaceUsesWith(val);
-                                    next = bb->remove(ip);
-                                }
-                            });
-                            if (!aval.isSingleValue() && !aval.isUnknown()) {
-                                auto phi = new Phi;
+                        // This load can have multiple values. We need a phi
+                        // to distinguish them. (LdFun case is not yet
+                        // handled here)
+                        if (!aval.isSingleValue() && !aval.isUnknown() &&
+                            !ldfun) {
+                            auto hasAllInputs = [&](BB* load) {
+                                bool success = true;
                                 aval.eachSource([&](ValOrig& src) {
```

Comment:
regarding inheriting from pair: yes, but it would make naming the fields more cumbersome (ie. would need functions, which means additional `()` for the clients. not sure what's better